### DISCUSSION TITLE
0.5.0 Released - with Pi Pico W support

### DISCUSSION THREAD

**Discussion by andrewdavidmackenzie** [andrewdavidmackenzie/pigg#656] [Announcements]
We're pleased to announce the [0.5.0 release](https://github.com/andrewdavidmackenzie/pigg/releases/tag/0.5.0) of "pigg"!!

0.5.0 brings many additions and improvements, the biggest one being initial support for Pi Pico W microcontrollers!
There is a UF2 firmware file as an asset in the release, which you can download and use to program your Pi Pico W!
Instructions on how to do that are included in the docs.

Consult the [0.5.0 release notes]((https://github.com/andrewdavidmackenzie/pigg/releases/tag/0.5.0)) for more details and let us know questions or problems via [GH Discussions](https://github.com/andrewdavidmackenzie/pigg/discussions) and [Issues](https://github.com/andrewdavidmackenzie/pigg/issues)

We are very interested in hearing from:
- Windows users who can help us test and verify "piggui" in particular
- Pi Pico W users
- Users in general, to guide future work

**Comment by DeeJay** [comment]
Initial feedback on v0.5.0 using an RPi5/RasPiOS as the workstation.

Summary - piggui on the workstation looks OK. Unable to use porky on an RPi Pico W.


piggui v0.5.0 installed on an RPi5 running 64-bit RasPiOS Bookworm. Installation of pigg via the 'one-line-installer.'

pre-built porky uf2 from releases/assets  installed on an RPi Pico W.

'Hardware' menu in piggui reports "Discovered devices (none)"

So unable to access the PicoW to configure wlan interface.



```
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ ls -al porky_pico_w.uf2 
-rw-r--r-- 1 david david 834048 Nov 19 23:30 porky_pico_w.uf2
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ file porky_pico_w.uf2 
porky_pico_w.uf2: UF2 firmware image, family Raspberry Pi RP2040, address 0x10000000, 1629 total blocks
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ picotool info porky_pico_w.uf2 
File porky_pico_w.uf2:

Program Information
 none
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ dmesg | tail -7
[343356.107344] usb 3-2: new full-speed USB device number 7 using xhci-hcd
[343356.265876] usb 3-2: not running at top speed; connect to a high speed hub
[343356.283878] usb 3-2: New USB device found, idVendor=babe, idProduct=face, bcdDevice= 0.10
[343356.283882] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[343356.283885] usb 3-2: Product: porky
[343356.283888] usb 3-2: Manufacturer: pigg
[343356.283890] usb 3-2: SerialNumber: e6616407e346bc28
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ lsusb
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 007: ID babe:face pigg porky
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ ls -al ~/.cargo/bin/piggui
-rwxr-xr-x 1 david david 13657232 Nov 19 23:21 /home/david/.cargo/bin/piggui
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ file ~/.cargo/bin/piggui
/home/david/.cargo/bin/piggui: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=d83bb3ff8da6935d7a17a867dcd78a2ad8c9e866, for GNU/Linux 3.7.0, not stripped
david@rp58book101064fnvme:~/Projects/pigg $ ~/.cargo/bin/piggui --version
piggui 0.5.0
david@rp58book101064fnvme:~/Projects/pigg $ 
david@rp58book101064fnvme:~/Projects/pigg $ 


```

**Comment by DeeJay** [comment]
Initial feedback on v0.5.0 using a Windows 10 Desktop as the workstation.

Summary - piggui on the workstation looks OK. Able to use porky on an RPi Pico W.

piggui v0.5.0 installed on an MS Windows10 desktop via pigg-x86_64-pc-windows-msvc

pre-built porky uf2 from releases/assets installed on an RPi Pico W.  Win10 Device Manager reports a usb device 'porky.'

Wlan configuration and subsequent connection works as documented, after some initial confusion and failure. (Didn't notice 'pop-UP' menu items amongst the gpio pin display.)

'Reset WLAN defaults' works, but piggui subsequently crashes/exits.


In use, Pico W pins GP0 and GP1 can ONLY be configured as OUTPUTS. All other pins can be INPUT or OUTPUT.

**Reply by andrewdavidmackenzie** [reply]
Thanks so much for testing it out and providing the feedback @DeeJay.

Special thanks for trying on Windows, you are the first one to do so, and to hear it works "OK", is great news!

Comments on points raised:
- after setting or resetting WLAN info, the device restarts so that it can try to connect to the default (there can be a default in ROM), not a crash - but it might appear as one. There is debug output explaining it will restart but not visible.
=> Maybe we could add something to the UI to let users know it will restart?
=> Maybe there is some USB status/command that can be used to make the restart not appear as a crash?

See: https://github.com/andrewdavidmackenzie/pigg/issues/659 I just created

- Confusion and failure: If you can elaborate more, maybe we can improve the docs or UI to reduce for others?

- GP0 and GP1 - This maybe a bug, left over from confusion on Pico W between GP0 and WL_GP0 etc .... I will file an issue and remedy.

See: https://github.com/andrewdavidmackenzie/pigg/issues/660 I just created.

If you have hardware that needs GP0 and GP1 now, let me know and I could push out a patch release to that.

**Reply by DeeJay** [reply]
_"If you have hardware that needs GP0 and GP1 now, let me know and I could push out a patch release to that."_

No urgency. Just observational feedback.

**Reply by DeeJay** [reply]
_"Confusion and failure: If you can elaborate more, maybe we can improve the docs or UI to reduce for others?"_

The Discovered Devices submenu is at the bottom of the main screen. Following that menu to the right shows the device details and serial  number. Reading further to the right the inline option is 'Reset'. I thought other options might appear further to the right. It simply didn't occur to me that there was a 'Drop-Up' menu that gave more options higher up the screen. Not sure how this might be clarified. It is trivial to use once it has been noticed. As discussed elsewhere, clicking on the Reset option is fatal in my experience.

**Comment by DeeJay** [comment]
Further feedback on v0.5.0 using an RPi5/RasPiOS as the workstation.


Having done the wireless configuration of a Pico W using an MS Windows version of piggui, I can subsequently connect to that networked porky instance from piggui running under  RasPiOS and basic trivial testing of toggling inputs and outputs (other than GP0 and GP1) works as expected.

So it is just the initial usb device discovery prior to wlan configuration that is the outstanding issue for me atm in respect of the RPi5/RasPiOS platform.

**Comment by DeeJay** [comment]
Tested a customised build of 'porky' uf2 to pre-configure default SSID.

Ref: **Customizing the default SSID information** [https://github.com/andrewdavidmackenzie/pigg/blob/0.5.0/porky/BUILDING.md#customizing-the-default-ssid-information](https://github.com/andrewdavidmackenzie/pigg/blob/0.5.0/porky/BUILDING.md#customizing-the-default-ssid-information)


Can confirm that it works as documented. There is a trivial typo concerning the` proky` project directory.

Actual documentation for "This uses the elf2usb2 command which you can install using cargo." would save some time and frustration.
Is `cargo install elf2uf2-rs
` the correct solution?


Build environment was 64-bit RasPiOS Bookworm on an RPi[5] board. Initial build attempts using a rust/cargo installation that was several months old failed. Had to refresh the build toolchain with `rustup upgrade` because rustc v.1.79.0 was not compatible.

RPI Pico W flashed with this .uf2 to create a 'porky' instance can be connected-to by piggui running on RasPiOS/RPi[5] and also from MS Windows Desktop. No further config needed beyond the IP:port combination.

**Reply by andrewdavidmackenzie** [reply]
Thanks for more testing and feedback!

- Typo being fixed in a PR. Thanks.
- Adding docs on how to install elf2uf2-rs in a PR. Thanks.
- On rust minimum version. That was the build of porky, right? I specify rust version of 1.80 in Cargo.toml for piggui, but now notice that I don't specify any for porky. Fixing in a PR. Good catch.

I have to say it's a testament to rust's cross-platform chops that all the host stuff works for you on Windows, and I have never compiled, tested or run it on Windows!

Send me some screen shots on Windows when you can please!

**Reply by andrewdavidmackenzie** [reply]
As an aside, I had considered building the porky binary (400K) into piggui, and allowing the host to flash devices directly. But I think as I will add Pi Pico (not W) and Pico2 and Pico2W over time, having so many options is not scalable.

Maybe it could pull the corresponding binary from the corresponding github release for you for the type of device detected....would you use that?

**Reply by andrewdavidmackenzie** [reply]
https://github.com/andrewdavidmackenzie/pigg/pull/662

**Reply by DeeJay** [reply]
_"Send me some screen shots on Windows when you can please!"_

As requested.

1. piggui on Windows 10, showing default 'fake' hardware interface. 
![Screenshot (1)](https://github.com/user-attachments/assets/7ab62266-35b9-4113-bd08-70fc4bbb4e62)
 
2. piggui on Windows 10, showing a piglet instance on an RPi[5] running RasPiOS Bookworm. 
![Screenshot (3)](https://github.com/user-attachments/assets/bc31cc67-dc5a-4e8a-b874-e84eb536b7bf)

3. Piggui on Windows 10, showing a porky instance on an RPi Pico W. 
![Screenshot (2)](https://github.com/user-attachments/assets/d69035c9-a247-48cc-98dd-fbf0459088f7)

**Reply by andrewdavidmackenzie** [reply]
Thanks for that. Kinda funny to see your app on a new platform for the first time! :-)

**Reply by andrewdavidmackenzie** [reply]
Is it OK if I add them (clipped) as images to our repo, for use in docs files, posts etc?

**Reply by DeeJay** [reply]
_"Is it OK if I add them [Windows screenshots] (clipped) as images to our repo, for use in docs files, posts etc?"_

OK to reuse them in support of the pigg project. Thanks for asking.

**Comment by DeeJay** [comment]
Posted initially for Discussion, but I think it might qualify as a Bug if you want to promote it to Issues...

piggui for Windows, with porky on Pico W: top part of display not visible in either Physical or Logical layouts.


The timeline/histogram display for GPI0 does not display the On/High/True value of the Pin.

![Screenshot (4) crop](https://github.com/user-attachments/assets/dead2f2d-e1ec-4a64-8c29-265a49b7b32d)

![Screenshot (5) crop](https://github.com/user-attachments/assets/dc273ad8-de5a-4831-98b9-679890063b8d)

**Reply by andrewdavidmackenzie** [reply]
Yes, a bug. There are small layout differences between mac and linux also that we avoid, sometimes due to default font sizes...

It looks pretty close, so probably needs just a few more pixels.

What about the bottom, is it OK?

I created this issue: https://github.com/andrewdavidmackenzie/pigg/issues/663

Is that correct?

I might need your help to confirm the fix, without going for a full-release in order to do so.
Can you build piggui from source if required?
If not, I'll investigate alternative methods...

**Reply by DeeJay** [reply]
_"What about the bottom, is it OK?"_

GP15/16 are OK in Board Layout mode.

In Logical Mode, something odd happens with GP28.

Setting GP28 alone to Output causes the High level of the trace to not be displayed. The Low level, adjacent to the edge of the Window, is fine.

If I set GP27 to Output at the same time then both High/Low of GP28 are logged correctly, independent of the state of GP27.

While testing this I noticed there is a similar interaction between GP0 and GP1 at the top of the screen. And at some stage while setting Statuses the piggui app on Windows crashed.

**Reply by DeeJay** [reply]
_"Can you build piggui from source if required?"_

I have only experimented with the rust toolchain under RasPiOS. If you don't have a Windows box for testing I assume some sort of cross-compile to an x86 .exe is possible? I was able to build the porky .uf2 under RasPiOS

**Reply by andrewdavidmackenzie** [reply]
Could you add some additional screenshots and descriptions to the issue?
I suspect it maybe caused by lack of vertical height in general, but will check.

You can try "make run-release" in root folder to see if piggui builds and runs OK. 
Will take a while the first time.

**Reply by DeeJay** [reply]
I am working on getting screenshots to add to #663.

Further investigation shows it is more widespread than just the top of the screen. 


To test a fix will need the ability to build an MS Windows .exe of piggui. I can build from source on RasPiOS linux for execution on RasPiOS. I can't see anything in the Docs (or in the Makefile) that would allow me to build ON RaspiOS for execution under x86_64 Windows?

**Reply by andrewdavidmackenzie** [reply]
Do you have a working rust tool chain on your windows machine?

If so, it would fall under the "build for your linux/mac/windows host" area....

you would just need to clone the repo, and then "cargo build", or using the Makefile "make run-release"... that will build and run a release profile version of "piggui", which will be a .exe if running on Windows.

**Reply by DeeJay** [reply]
For the time being the conclusion is that I will not be able to test a fix for Windows if it requires building from source.

**Comment by andrewdavidmackenzie** [comment]
@DeeJay I-ve installed Parallels on my Mac with a copy of WIndows 11 - to see if I can reproduce, debug and develop piggui on WIndows and fix the issues you reported.

**Comment by andrewdavidmackenzie** [comment]
I think this error dialog has put an end to that experiment, but will search to make sure...

<img width="696" alt="Screenshot 2024-12-15 at 11 59 15 AM" src="https://github.com/user-attachments/assets/e0529ab0-ac4b-4880-81ae-7ced2d4671e5" />

**Comment by andrewdavidmackenzie** [comment]
Well, to my own surprise, I was able to overcome that. I think it's compiling and building for x86 on windows, using ARM emulation layer, inside Parallels, on my Mac.... 

Either way, it-s working...

<img width="2107" alt="Screenshot 2024-12-17 at 9 37 39 AM" src="https://github.com/user-attachments/assets/2a1ac626-9ae9-4bb3-9336-6380d0147d77" />

So, now I will try to reproduce your errors. 

I'll provide updates in the issue -  https://github.com/andrewdavidmackenzie/pigg/issues/663

**Reply by andrewdavidmackenzie** [reply]
@DeeJay pushed a PR "with fixes". Not sure if it will fix this though as I couldn't reproduce (on Windows 11).
Possibly for you to build HEAD of master and try it?
If not, then we can see if fixed by upcoming 0.6.0 release...
